lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -225,56 +225,40 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
-# Collect text output into JSON. save_results() is called too early for this.
+# incorporate textual step data into details JSON
+# note: Can not be called from save_results() because the upload must have already been concluded.
 sub finalize_results {
     my ($self) = @_;
 
-    my $dir = $self->job->result_dir();
-    return 0 unless $dir;
-    my $file    = path($dir, "details-" . $self->name . ".json");
-    my $tmpfile = $file->sibling($file->basename . '.tmp');
-    return 0 unless -e $file;
-    my $errors = 0;
-    my $results;
-
-    eval {
-        $results = decode_json($file->slurp);
-        $results = {details => $results} if ref($results) eq 'ARRAY';
-
-        for my $step (@{$results->{details}}) {
-            next unless $step->{text};
-            my $txtfile = path($dir, $step->{text});
-            next unless -e $txtfile;
-            $step->{text_data} = decode('UTF-8', $txtfile->slurp);
-        }
+    # locate details JSON; skip if not present
+    my $dir = $self->job->result_dir;
+    return undef unless $dir;
+    my $name = $self->name;
+    my $file = path($dir, "details-$name.json");
+    return undef unless -e $file;
 
-        $tmpfile->spurt(encode_json($results));
-
-        rename $tmpfile->to_string, $file->to_string
-          or die "finalize_results(" . $self->job->id . "): Cannot update details-" . $self->name . ".json. $!";
-    };
-
-    if (my $err = $@) {
-        log_error $err;
-        $errors = 1;
+    # read details; skip if none present
+    my $results = decode_json($file->slurp);
+    my $details = ref $results eq 'HASH' ? $results->{details} : $results;
+    return undef unless ref $details eq 'ARRAY' && @$details;
+
+    # incorporate textual step data into details
+    for my $step (@$details) {
+        next unless my $text = $step->{text};
+        my $txtfile = path($dir, $text);
+        next unless -e $txtfile;
+        $step->{text_data} = decode('UTF-8', $txtfile->slurp);
     }
 
-    eval { $tmpfile->remove if -e $tmpfile; };
-    return !$errors;
-}
-
-# Delete result files which were merged into JSON details file.
-sub cleanup_results {
-    my ($self) = @_;
-
-    my $dir = $self->job->result_dir();
-    return unless $dir;
-    my $file = path($dir, "details-" . $self->name . ".json");
-    return unless -e $file;
-    my $results = decode_json($file->slurp);
+    # replace file contents on disk using a temp file to preserve old file if something goes wrong
+    my $new_file_contents = encode_json($results);
+    my $tmpfile           = tempfile(DIR => $file->dirname);
+    $tmpfile->spurt($new_file_contents);
+    $tmpfile->move_to($file);
```

Comment:
Tempfiles are created with access mode `0600`. WebUI backend won't be able to read the JSON files if the Gru service runs under a different system user.

Reply:
Ok, that explains why your version "awkwardly" didn't use tempfile. I haven't thought about this because all daemons are supposed to run under the same user anyways. However, it might nevertheless be a good idea to allow global reading.